File: tests/layer0_ops/unit/test_signal_extension.py
"""
test_signal_extension.py
========================
Python port of:
    layer0_ops/unit/TestSignalExtension.m

Covers: comp_extBoundary
"""

from __future__ import annotations

import pytest

import numpy as np
from conftest import comp_extBoundary_ref

_ALL_MODES = ("per", "sym", "symw", "asym", "asymw", "sp0", "zpd")


# ---------------------------------------------------------------------------
# comp_extBoundary – reference tests
# ---------------------------------------------------------------------------

class TestCompExtBoundaryReference:
    """
    MATLAB counterpart: TestSignalExtension.
    """

    def test_periodic_left_extension(self):
        """'per' left extension == tail of f."""
        rng    = np.random.default_rng(42)
        f      = rng.standard_normal(64)
        extLen = 16
        f_ext  = comp_extBoundary_ref(f, extLen, "per")

        np.testing.assert_allclose(
            f_ext[:extLen], f[len(f) - extLen:],
            atol=1e-14,
            err_msg="Periodic left extension does not match tail of f"
        )

    def test_periodic_right_extension(self):
        """'per' right extension == head of f."""
        rng    = np.random.default_rng(42)
        f      = rng.standard_normal(64)
        extLen = 16
        L      = len(f)
        f_ext  = comp_extBoundary_ref(f, extLen, "per")

        np.testing.assert_allclose(
            f_ext[L + extLen:],
            f[:extLen],
            atol=1e-14,
            err_msg="Periodic right extension does not match head of f"
        )

    def test_zpd_left_zeros(self):
        """'zpd' left extension is zero."""
        rng    = np.random.default_rng(42)
        f      = rng.standard_normal(64)
        extLen = 16
        f_ext  = comp_extBoundary_ref(f, extLen, "zpd")

        np.testing.assert_allclose(f_ext[:extLen], np.zeros(extLen), atol=1e-14)

    def test_zpd_right_zeros(self):
        """'zpd' right extension is zero."""
        rng    = np.random.default_rng(42)
        f      = rng.standard_normal(64)
        extLen = 16
        L      = len(f)
        f_ext  = comp_extBoundary_ref(f, extLen, "zpd")

        np.testing.assert_allclose(f_ext[L + extLen:], np.zeros(extLen), atol=1e-14)

    def test_sym_output_length(self):
        """'sym' mode: total length == L + 2*extLen."""
        rng    = np.random.default_rng(42)
        f      = rng.standard_normal(64)
        extLen = 16
        f_ext  = comp_extBoundary_ref(f, extLen, "sym")

        assert len(f_ext) == len(f) + 2 * extLen

    @pytest.mark.parametrize("mode", _ALL_MODES)
    def test_all_modes_total_length(self, mode):
        """All modes: total length == L + 2*extLen."""
        rng    = np.random.default_rng(42)
        f      = rng.standard_normal(64)
        extLen = 16
        f_ext  = comp_extBoundary_ref(f, extLen, mode)

        assert len(f_ext) == len(f) + 2 * extLen, \
            f"Mode {mode}: expected {len(f) + 2*extLen}, got {len(f_ext)}"

    def test_zero_extlen_identity(self):
        """extLen=0: output identical to input."""
        rng   = np.random.default_rng(42)
        f     = rng.standard_normal(64)
        f_ext = comp_extBoundary_ref(f, 0, "per")

        assert len(f_ext) == len(f)
        np.testing.assert_allclose(f_ext, f, atol=1e-14)

    @pytest.mark.parametrize("mode", ("per", "sym", "zpd"))
    def test_original_signal_preserved(self, mode):
        """Middle segment f_ext[extLen : extLen+L] == f for common modes."""
        rng    = np.random.default_rng(42)
        L      = 128
        f      = rng.standard_normal(L)
        extLen = L // 4
        f_ext  = comp_extBoundary_ref(f, extLen, mode)

        assert len(f_ext) == L + 2 * extLen
        np.testing.assert_allclose(
            f_ext[extLen: extLen + L], f, atol=1e-14,
            err_msg=f"Mode {mode}: original signal not preserved in the middle"
        )

    def test_extlen_equals_L(self):
        """extLen=L: total output length == 3*L."""
        rng    = np.random.default_rng(42)
        L      = 64
        f      = rng.standard_normal(L)
        f_ext  = comp_extBoundary_ref(f, L, "per")

        assert len(f_ext) == 3 * L

    @pytest.mark.parametrize("mode", _ALL_MODES)
    def test_asymmetric_signal_all_modes(self, mode):
        """All modes produce correct length for a generic random signal."""
        rng    = np.random.default_rng(42)
        f      = rng.standard_normal(64)
        extLen = 8
        f_ext  = comp_extBoundary_ref(f, extLen, mode)

        assert len(f_ext) == 64 + 2 * 8, \
            f"Mode {mode}: length mismatch"


# ---------------------------------------------------------------------------
# comp_extBoundary – implementation tests
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestCompExtBoundaryImpl:
    """Verify cool_frames.layer0.comp_extBoundary matches reference."""

    @pytest.mark.parametrize("mode", ("per", "zpd", "sym"))
    def test_matches_reference(self, needs_impl, mode):
        from cool_frames.core import comp_extBoundary  # type: ignore

        rng = np.random.default_rng(42)
        for _ in range(5):
            L      = int(rng.integers(32, 128))
            f      = rng.standard_normal(L)
            extLen = int(rng.integers(4, L // 4 + 1))
            np.testing.assert_allclose(
                comp_extBoundary(f, extLen, mode),
                comp_extBoundary_ref(f, extLen, mode),
                atol=1e-14,
                err_msg=f"mode={mode}"
            )

File: tests/layer0_ops/unit/TestSignalExtension.m
classdef TestSignalExtension < matlab.unittest.TestCase
    % TestSignalExtension: unit tests for comp_extBoundary

    methods(TestClassSetup)
        function setupTestClass(testCase)
            sharedDir = fullfile(fileparts(mfilename('fullpath')), ...
                                 '..', '..', 'shared');
            addpath(sharedDir);
            setup_filterbank_paths();
        end
    end

    methods(Test)

        function testExtBoundaryPeriodicMode(testCase)
            % 'per' mode: layout is [f(end-extLen+1:end) ; f ; f(1:extLen)]
            % Left extension  = tail of f,  right extension = head of f.
            rng(42);
            f = randn(64, 1);
            extLen = 16;
            f_ext = comp_extBoundary(f, extLen, 'per');

            L = length(f);
            % Left extension (positions 1:extLen) == last extLen samples of f
            testCase.verifyEqual(f_ext(1:extLen), f(L-extLen+1:L), 'AbsTol', 1e-14, ...
                'Periodic left extension does not match tail of f');
            % Right extension (positions L+extLen+1:end) == first extLen samples of f
            testCase.verifyEqual(f_ext(L+extLen+1:end), f(1:extLen), 'AbsTol', 1e-14, ...
                'Periodic right extension does not match head of f');
        end

        function testExtBoundaryZpdMode(testCase)
            % 'zpd' mode: extended region is zero
            rng(42);
            f = randn(64, 1);
            extLen = 16;
            f_ext = comp_extBoundary(f, extLen, 'zpd');

            L = length(f);
            testCase.verifyEqual(f_ext(1:extLen), zeros(extLen, 1), 'AbsTol', 1e-14);
            testCase.verifyEqual(f_ext(L+extLen+1:end), zeros(extLen, 1), 'AbsTol', 1e-14);
        end

        function testExtBoundarySymMode(testCase)
            % 'sym' mode: correct length
            rng(42);
            f = randn(64, 1);
            extLen = 16;
            f_ext = comp_extBoundary(f, extLen, 'sym');

            L = length(f);
            testCase.verifyEqual(length(f_ext), L + 2*extLen);
        end

        function testExtBoundaryLength(testCase)
            % Extension has correct total length: L + 2*extLen for all modes
            rng(42);
            f = randn(64, 1);
            modes = {'per', 'sym', 'symw', 'asym', 'asymw', 'sp0', 'zpd'};

            for mode = modes
                extLen = 16;
                f_ext = comp_extBoundary(f, extLen, mode{1});
                expected_len = length(f) + 2*extLen;
                testCase.verifyEqual(length(f_ext), expected_len, ...
                    sprintf('Mode %s: incorrect length', mode{1}));
            end
        end

        function testExtBoundaryZeroExtension(testCase)
            % extLen = 0: output identical to input
            rng(42);
            f = randn(64, 1);
            extLen = 0;
            f_ext = comp_extBoundary(f, extLen, 'per');

            testCase.verifyEqual(length(f_ext), length(f));
            testCase.verifyEqual(f_ext, f, 'AbsTol', 1e-14);
        end

        function testExtBoundaryOriginalPreserved(testCase)
            % Middle section (extLen+1 : extLen+L) always equals the original f
            rng(42);
            L = 128;
            f = randn(L, 1);
            extLen = L / 4;

            modes = {'per', 'sym', 'zpd'};
            for mode = modes
                f_ext = comp_extBoundary(f, extLen, mode{1});
                testCase.verifyEqual(length(f_ext), L + 2*extLen);
                testCase.verifyEqual(f_ext(extLen+1:extLen+L), f, 'AbsTol', 1e-14, ...
                    sprintf('Mode %s: original signal not preserved in the middle', mode{1}));
            end
        end

        function testExtBoundaryFullLength(testCase)
            % extLen = L: total output length = 3*L
            rng(42);
            L = 64;
            f = randn(L, 1);
            extLen = L;

            f_ext = comp_extBoundary(f, extLen, 'per');
            testCase.verifyEqual(length(f_ext), 3*L);
        end

        function testExtBoundaryAsymmetricSignal(testCase)
            % All modes work without error on a generic random signal
            rng(42);
            f = randn(64, 1);
            extLen = 8;

            modes = {'per', 'sym', 'symw', 'asym', 'asymw', 'sp0', 'zpd'};
            for mode = modes
                f_ext = comp_extBoundary(f, extLen, mode{1});
                testCase.verifyEqual(length(f_ext), 64 + 2*8, ...
                    sprintf('Mode %s: length mismatch', mode{1}));
            end
        end

        function testExtBoundaryAllModes(testCase)
            % All named modes run without exception
            rng(42);
            f = randn(64, 1);
            extLen = 8;

            modes = {'per', 'sym', 'symw', 'asym', 'asymw', 'sp0', 'zpd'};
            for mode = modes
                try
                    f_ext = comp_extBoundary(f, extLen, mode{1});
                    testCase.verifyEqual(length(f_ext), 64 + 2*8);
                catch ME
                    testCase.verifyFail(sprintf('Mode %s failed: %s', mode{1}, ME.message));
                end
            end
        end

    end

end
